---
title: "Generate gff and fasta files combining the reference S288C genome with plasmid construct sequences"
author: "Sam Haynes"
date: "02/07/2021"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(here)

experiment_name = "QuantSeq_heat_shock" # folder name for experiential 
                             # meta either QuantSeq or 5PSeq

GTF_COLNAMES = c("seqname", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")
FEATURES = c("primary_transcript","Promoter", "CDS", "Terminator", "URA3", "M1", "M2", "M3")
```

```{r function}
write_gff <- function(construct_annotation_tibble, genome_annotation_tibble){
  sample_name <- construct_annotation_tibble %>%
    pull(sample_name) %>%
    unique()
  
  construct_name <- construct_annotation_tibble %>%
    pull(seqname) %>%
    unique()
  
  full_annotation_tibble <- construct_annotation_tibble %>% 
    ungroup() %>%
    select(seqname, source, feature, start, end, score, strand, frame, attribute) %>%
    bind_rows(genome_annotation_tibble)
    
  output_gff_file <- here(paste0("./data/",
                                 experiment_name,
                                 "/input/genome_annotations/", 
                                 sample_name,
                                 "_sample_longest_full_ORF_with_constructs.gff"))

write_lines(
    x=c("##gff-version 3", 
        paste0("##", output_gff_file),
        "##source-version yeastutrgff",
        paste("##date",lubridate::today())),
    path=output_gff_file)

write_tsv(full_annotation_tibble, output_gff_file, append = TRUE)
}

write_fasta <- function(construct_data, genomic_fasta){
  construct_fasta_file <- construct_data$filepath
  
  construct_name <- construct_data$construct_name
  
  sample_name <- construct_data$sample_name
  
  construct_fasta <- read_table(construct_fasta_file)
  
  construct_genome_fasta_file_path <- here(paste0("./data/",
                                                  experiment_name,
                                                  "/input/genome_fasta/",
                                                  sample_name,
                                                  "_sample_with_saccharomyces_cerevisiae_R64.fa"))
  
  write_tsv(genomic_fasta, 
            path = construct_genome_fasta_file_path, 
            col_names = TRUE)

write_tsv(construct_fasta,
  path = construct_genome_fasta_file_path,
  append=TRUE,
  col_names = TRUE)
}
```

```{r generate-plasmid-feature-annotations}
plasmid_features <- read_delim(here("./data/shared_data/raw_annotation/construct_annotation/plasmid_features.csv"), ";")

RNA_seq_sample_names <- read_csv(here(paste0("./data/", 
                                             experiment_name,
                                             "/input/experiment_metadata/",
                                             "sample_to_construct_table.csv")))

plasmid_features_long <- plasmid_features %>% 
  select(-Linear_sequence,-Orig_Promoter,-Orig_Terminator) %>% 
  pivot_longer(!Construct_name,names_to = "feature", values_to = "position") %>% 
  separate(feature,into=c("feature","location"),sep="_") %>% 
  pivot_wider(names_from = location, values_from = position) %>%
  filter(!is.na(start))

plasmid_primary_transcript_feature <- plasmid_features_long %>%
  filter(feature != "URA3") %>%
  group_by(Construct_name) %>%
  summarise(feature = "primary_transcript", start = min(start), stop = max(stop))

plasmid_features_no_URA3_primary_transcript <- plasmid_features_long %>%
  bind_rows(plasmid_primary_transcript_feature) %>%
  mutate(strand = ifelse(feature == "URA3", "-", "+")) %>%
  mutate(source = "chimera-project", score = ".", frame = ".", attribute = paste0("ID=", Construct_name, "_", feature, ";Name=", Construct_name, "_", feature), seqname = Construct_name, end = stop) %>%
  select(GTF_COLNAMES)

# Adds a URA3 primary_transcript feature to enable counting with featureCounts

plasmid_features_all <- plasmid_features_no_URA3_primary_transcript %>%
  bind_rows(plasmid_features_long %>%
              filter(feature == "URA3") %>%
              mutate(feature = "primary_transcript", 
                     strand = "-", 
                     source = "chimera-project", 
                     score = ".", 
                     frame = ".", 
                     attribute = paste0("ID=", Construct_name, "_URA3", ";Name=", Construct_name, "_URA3"), 
                     seqname = Construct_name, 
                     end = stop) %>%
              select(GTF_COLNAMES)
  )

plasmid_features_all_two_reps <- plasmid_features_all %>%
  mutate(rep = paste0(seqname, "_b1")) %>%
  bind_rows(plasmid_features_all %>%
              mutate(rep = paste0(seqname, "_b2"))) %>%
  inner_join(RNA_seq_sample_names, 
             by = c("rep" = "construct_name"))

genome_features <- read_tsv(here("data/shared_data/raw_annotation/reference_genome_annotation/longest_full-ORF_ypd_plus_other_fixed_UTR_length_transcripts.gff"), 
                            comment="#", 
                            col_names = GTF_COLNAMES) %>%
  mutate(attribute = paste0(attribute, 
                            ";Name=", 
                            str_extract(attribute, "(?<=ID\\=)\\w+")))

# Write separate gff files for each sample
plasmid_features_all_two_reps %>%
  group_by(sample_name) %>%
  do(write_gff(.,genome_features))

```

```{r create-plasmid-fasta-files}
construct_fasta_file_names <- tibble(filepath = list.files(here("./data/shared_data/raw_fasta/construct_fasta"), full.names = TRUE)) %>%
  mutate(construct_name = str_extract(filepath, "(?<=linear_)[a-z0-9A-Z\\-_]+(?=.fa)"))

construct_fasta_file_names_two_reps <- construct_fasta_file_names %>%
  mutate(construct_name = paste0(construct_name, "_b1")) %>%
  bind_rows(construct_fasta_file_names%>%
              mutate(construct_name = paste0(construct_name, "_b2"))) %>%
  inner_join(RNA_seq_sample_names)

# Remove URA3 from genomic fasta file
bedtoolsr::bt.maskfasta(here("./data/shared_data/raw_fasta/reference_genome_fasta/saccharomyces_cerevisiae_R64-2-1_20150113_chrnames.fsa"), here("./data/shared_data/raw_fasta/reference_genome_fasta/saccharomyces_cerevisiae_R64_delta_URA3.fa"), here("./data/shared_data/raw_annotation/reference_genome_annotation/URA3_deletion_position.bed"))

# read in genomic fasta file
genomic_fasta <- read_table(here("./data/shared_data/raw_fasta/reference_genome_fasta/saccharomyces_cerevisiae_R64_delta_URA3.fa"))

construct_fasta_file_names_two_reps %>%
  group_by(sample_name) %>%
  do(write_fasta(.,genomic_fasta))

```
